In [1]:
# Import the necessary libraries 
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras.preprocessing.image  import ImageDataGenerator

In [2]:
#Import the dataset 
dataset_path=r"C:\Users\aksha\Desktop\AI&DS TRAINING\CNN\PetImages"


In [3]:
# Image prprocessing
datagen=ImageDataGenerator(
    #Normalize pixels 
    rescale=1./255,
    # Data augementation
    shear_range=0.2,
    zoom_range=0.2,
    horizontal_flip=True,

    # Split dataset 
    validation_split=0.2 # Validation split used to create train and validation data
)

In [4]:
# Load traing data 
# Training data 
train_Data=datagen.flow_from_directory(
    dataset_path,
    target_size=(128,128),
    batch_size=32,
    class_mode='binary',
    subset='training'
)

Found 12947 images belonging to 2 classes.


In [5]:
# load validation data 
val_data=datagen.flow_from_directory(
    dataset_path,
    target_size=(128,128),
    batch_size=32,
    class_mode='binary',
    subset='validation'
)

Found 3235 images belonging to 2 classes.


In [6]:
print(train_Data.class_indices)

{'Cat': 0, 'Dog': 1}


In [7]:
# Building CNN model 
model=keras.Sequential()

In [8]:
# 1.Simple Images ---> Few Layers 
# ex: handwritten numbers
# (28*28)
# 2 conc 2 pool
 
# 2.Medium Complexity
# Cats vs Dogs
# (128*128)
# (224*224)
# 3 Conv + 3 Pool.

In [9]:
# First convolutional layer 
model.add(
    keras.layers.Conv2D(
        filters=32,
        kernel_size=(3,3),
        activation='relu',
        input_shape=(128,128,3)
    )
)


c:\Users\aksha\Documents\anaconda\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:113: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [10]:
# First pooling layer 
model.add(
    keras.layers.MaxPooling2D(
        pool_size=(2,2)
        )
)

In [11]:
# 2 convolutional layer 
model.add(
    keras.layers.Conv2D(
        filters=32,
        kernel_size=(3,3),
        activation='relu'
    )
)

In [12]:
# Second pooling layer 
model.add(
    keras.layers.MaxPooling2D(
        pool_size=(2,2)
    )
)

In [13]:
# Flatten 
model.add(keras.layers.Flatten())

In [14]:
# Dense layer 
model.add(
    keras.layers.Dense(
        units=128,
        activation='relu'
        )
)

In [15]:
# Output layer 
model.add(
    keras.layers.Dense(
        units=1,
        activation='sigmoid'
    )
)

In [16]:
# view  Model architecture 
model.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d (Conv2D)                 │ (None, 126, 126, 32)   │           896 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d (MaxPooling2D)    │ (None, 63, 63, 32)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_1 (Conv2D)               │ (None, 61, 61, 32)     │         9,248 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_1 (MaxPooling2D)  │ (None, 30, 30, 32)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten (Flatten)               │ (None, 28800)          │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 128)            │     3,686,528 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 1)              │           129 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 3,696,801 (14.10 MB)

 Trainable params: 3,696,801 (14.10 MB)

 Non-trainable params: 0 (0.00 B)

(None,126,126,Channels)

(Batch size,Height,Width,No.of filters)

#### 1 Conv2D

input=128*128*3 

filter=3*3

no. of filters=32 

output=((Input-kernel)/stride)+1 

Stride=1 

output=((128-3)/1)+1
       =126 

height: 126 

width:126 

Channels--->32 

Final output --->(None,126,126,32)

kernel 

Formula=(kernel height* kernel width* Input channels+1)*Filters

kernel=3*3 

Inputchannels=3

filters=32 

params=(3*3*3+1)*32

        =(27+1)*32 

        =28*32 

        =896 

#### First Pooling 2D 

MaxPooling2D((2,2))

Pooling= No learning,No weights 

(only reduces size) 

Input = 126*126*32 

pool=2*2

Formula=126/2 

       = 63

Output : 63*63*32  

Final:(None,63,63,32)

#### Second Conv2D

Input from pooling: (None,63* 63 *32)

kernel: 3*3 

output: 63-3+1

    = 61 

(None,61,61,32)

Params= 9248 

kernels=(3*3*32+1)*32 

       =(288+1)*32

       =289*32 

       =9248

#### Second Pooling 

(None,30,30,32)

Param=0 

Input=(61,61,32)

Param=0 

Pooling:2*2 

Floor division:61/2

               =30 

Final: (None,30,30,32)

No parameters.

#### Flatten 

Input from the second pooling layer:30*30 *32

Output: 30*30*32 =28800 

flatten: It will convert 3D--->1D vector (Flatten only reshapes...)

#### Dense Layer

Input: 28800

Neurons=128 

Parameters =(Input*Neurons)+Bias 

        =(28800*128)+128 

        =3686400+128

        =3686528

### Output layer 

Input=128

Output=1 

Output shape (None,1)

Meaning=Single prediction 

----->Cat or Dog 

Parameter calculation:

        =128*1+1

        =129


Input--->Conv1--->Pool1---->Conv2---->Pool2--->Flatten--->Dense--->Output


In [17]:
import tensorflow as tf 
print(tf.config.list_physical_devices())

[PhysicalDevice(name='/physical_device:CPU:0', device_type='CPU')]


In [18]:
# Compile the model 
model.compile(
    optimizer='adam',
    loss='binary_crossentropy',
    metrics=['accuracy']
)

In [21]:
# Train 
history=model.fit(
    train_Data,
    epochs=2,
    validation_data=val_data
)

Epoch 1/2
405/405 ━━━━━━━━━━━━━━━━━━━━ 139s 344ms/step - accuracy: 0.7973 - loss: 0.4623 - val_accuracy: 0.8031 - val_loss: 0.4484
Epoch 2/2
405/405 ━━━━━━━━━━━━━━━━━━━━ 141s 349ms/step - accuracy: 0.8122 - loss: 0.4326 - val_accuracy: 0.8179 - val_loss: 0.4179


In [22]:
# Evaluate the model 
loss,accuracy=model.evaluate(val_data)
print("Accuracy: ",accuracy)

102/102 ━━━━━━━━━━━━━━━━━━━━ 21s 204ms/step - accuracy: 0.8232 - loss: 0.4101
Accuracy:  0.8231839537620544
